File: noeyeyh/Week26/아이템 줍기.py
from collections import deque

def solution(rectangle, characterX, characterY, itemX, itemY):
    directions = [(-1, 0), (0, 1), (1, 0), (0, -1)]
    
    MAX = 101
    maps = [[0 for _ in range(MAX)] for _ in range(MAX)]
    visited = [[0 for _ in range(MAX)] for _ in range(MAX)]
    
    rectangle = [[x1 * 2, y1 * 2, x2 * 2, y2 * 2] for x1, y1, x2, y2 in rectangle]
    
    for x1, y1, x2, y2 in rectangle:
        for x in range(x1, x2 + 1):
            maps[x][y1] = 1
            maps[x][y2] = 1
        for y in range(y1, y2 + 1):
            maps[x1][y] = 1
            maps[x2][y] = 1

    for x1, y1, x2, y2 in rectangle:
        for i in range(MAX):
            for j in range(MAX):
                if maps[i][j] == 1 and x1 < i < x2 and y1 < j < y2:
                    maps[i][j] = 0

    queue = deque([(characterX*2, characterY*2, 0)])
    
    while queue:
        x, y, c = queue.popleft()
        
        if x == itemX*2 and y == itemY*2: return c//2
        
        for dx, dy in directions:
            nx = x + dx
            ny = y + dy
            
            if nx < 0 or nx >= MAX or ny < 0 or ny >= MAX or maps[nx][ny] == 0 or visited[nx][ny] == 1:
                continue
            visited[nx][ny] = 1
            queue.append((nx, ny, c + 1))
    
    return -1

File: un6unw/Week26/아이템 줍기.java
import java.util.*;

class Solution {
    public int solution(int[][] rectangle, int characterX, int characterY, int itemX, int itemY) {
        int[][] map = new int[102][102]; // 2배 확대하여 처리
        for (int[] rect : rectangle) {
            int x1 = rect[0] * 2, y1 = rect[1] * 2;
            int x2 = rect[2] * 2, y2 = rect[3] * 2;

            for (int i = x1; i <= x2; i++) {
                for (int j = y1; j <= y2; j++) {
                    if (i == x1 || i == x2 || j == y1 || j == y2) {
                        if (map[i][j] == 0) map[i][j] = 1; // 테두리
                    } else {
                        map[i][j] = 2; // 내부
                    }
                }
            }
        }

        // BFS 탐색
        Queue<int[]> queue = new LinkedList<>();
        boolean[][] visited = new boolean[102][102];
        queue.add(new int[]{characterX * 2, characterY * 2, 0});
        visited[characterX * 2][characterY * 2] = true;

        int[] dx = {1, -1, 0, 0};
        int[] dy = {0, 0, 1, -1};

        while (!queue.isEmpty()) {
            int[] cur = queue.poll();
            int x = cur[0], y = cur[1], dist = cur[2];

            if (x == itemX * 2 && y == itemY * 2) {
                return dist / 2; // 원래 크기로 되돌림
            }

            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];

                if (nx < 0 || ny < 0 || nx >= 102 || ny >= 102) continue;
                if (visited[nx][ny] || map[nx][ny] != 1) continue;

                visited[nx][ny] = true;
                queue.add(new int[]{nx, ny, dist + 1});
            }
        }

        return -1;
    }
}